src/net_processing.cpp

```diff
@@ -358,7 +358,7 @@ struct Peer {
     std::deque<CInv> m_getdata_requests GUARDED_BY(m_getdata_requests_mutex);
 
     /** Time of the last getheaders message to this peer */
-    std::atomic<std::chrono::seconds> m_last_getheaders_timestamp{0s};
+    std::atomic<NodeClock::time_point> m_last_getheaders_timestamp{NodeSeconds{}};
```

Comment:
Is there even a need to make this atomic? In a single threaded environment this is not needed. And if the same peer is served by several threads (for whatever reason) in the future, an atomic may prevent UB, but doesn't seem sufficient to prevent races logically.

Reply:
Hmm that is an interesting point.  My sense was that we don't want to rely on implicit single-threadedness to reason about data races, so it was safer to just throw this into an atomic and not worry about it.

I also think that having multiple threads servicing the same peer would not really work for a bunch of reasons (including the one you give, about logical errors in code like this).  I would imagine that multithreading network handling in the future would involve different threads servicing different peers.

I think the most likely form of UB from data races would be if we were to expose this value via RPC in the future.  So to make this future-proof, I'd prefer that we leave this as-is; it doesn't seem like the performance hit of having an atomic here (which is rarely accessed, as `getheaders` messages are not a frequent occurrence, and there are far more effective ways to CPU DoS a node than to trigger a getheaders) is significant enough to warrant making this harder to reason about in the future.

Reply:
I think if this was ever exposed on RPC, we'd also want to think which other fields to return atomically in the response, in which case a per-field atomic isn't enough to prevent logic races. So to me it seems slightly preferable to leave the design of thread safety to when it is needed. Otherwise it may be assumed in the future that this is already perfectly thread safe and can be used as is. Leaving this plain (without atomic) would at least have a thread sanitizer failure hint that the design needs to be re-thought if this was accessed by more than one thread in the future.

Reply:
(Other than that, agree that atomic doesn't affect performance)

Reply:
You know, I think you're right that this atomic is incorrect.  We need to hold a lock throughout `MaybeSendGetHeaders()` because this variable can be accessed in the scheduler thread or in the net processing thread, and if we don't hold a lock throughout that function we can get a data race.  Thanks for asking about this, I'll open a PR to fix.

Reply:
Addressed in #25557

Reply:
Can you elaborate a bit on this? Unless I am missing something `MaybeSendGetHeaders`/`ConsiderEviction` is only called in-thread, not to be confused with `CheckForStaleTipAndEvictPeers`, which is called out-of-thread. What am I missing?

Reply:
What is "in-thread" and "out-of-thread"?

Reply:
"in-thread" means running in the thread "ThreadMessageHandler (b-msghand): Application level message handling (sending and receiving). Almost all net_processing and validation logic runs on this thread."

Reply:
Right, `m_last_getheaders_timestamp` is accessed from `PeerManagerImpl::MaybeSendGetHeaders()` and `PeerManagerImpl::ProcessMessage()` which are only called from a single thread `ThreadMessageHandler()`.

> because this variable can be accessed in the scheduler thread or in the net processing thread

@sdaftuar, I think it is not accessed from the scheduler thread?

I am in favor of simplicity, in this case meaning to not protect data that is accessed by a single thread. If in the future it is accessed by another thread, then an appropriate protection should be added. That may be an atomic or a mutex just for that variable or a mutex for multiple variables depending on the need. Otherwise we may end up protecting it needlessly now and later to have to rework the protection anyway (or worse - use the available, inappropriate protection, like @MarcoFalke mentioned above "...it may be assumed in the future that this is already perfectly thread safe...").

Reply:
> Can you elaborate a bit on this? Unless I am missing something MaybeSendGetHeaders/ConsiderEviction is only called in-thread, not to be confused with CheckForStaleTipAndEvictPeers, which is called out-of-thread. What am I missing?

Ah, I guess I misremembered how `ConsiderEviction` is called (and thought it was run in the scheduler thread).  

I'll update my other PR and drop the mutex/atomic altogether.

Reply:
> I am in favor of simplicity, in this case meaning to not protect data that is accessed by a single thread.

Possible idea for the future. In theory could have `SingleThread<T>` class similar to the `Synced<T>` class from #25390 that wraps an object and asserts in debug mode that is always accessed from the same thread. I think a solution like that would be overkill here, and I'm struggling to think of cases where it wouldn't be overkill, but it wouldn't be hard to implement and could be useful to help code document assumptions it is making.

Reply:
Should be resolved by 8efa73e7ce4ea0c1b7ad5c3947a5ecf9fb6361d3